thevoid/connection.cpp

```diff
@@ -83,6 +84,97 @@ struct timespec gettime_now() {
 
 } // unnamed namespace
 
+
+namespace {
+
+template <typename InputIterator, typename OutputIterator>
+OutputIterator escape(
+		InputIterator begin, InputIterator end,
+		OutputIterator output,
+		int quote
+	)
+{
+	for (auto iter = begin; iter != end; ++iter) {
+		char ch = *iter;
+
+		if (ch == '\a') {
+			output++ = '\\'; output++ = 'a';
+		}
+		else if (ch == '\b') {
+			output++ = '\\'; output++ = 'b';
+		}
+		else if (ch == '\f') {
+			output++ = '\\'; output++ = 'f';
+		}
+		else if (ch == '\n') {
+			output++ = '\\'; output++ = 'n';
+		}
+		else if (ch == '\r') {
+			output++ = '\\'; output++ = 'r';
+		}
+		else if (ch == '\t') {
+			output++ = '\\'; output++ = 't';
+		}
+		else if (ch == '\v') {
+			output++ = '\\'; output++ = 'v';
+		}
+		else if (ch == quote || ch == '\\') {
+			output++ = '\\'; output++ = ch;
+		}
+		else if (ch < ' ' || ch >= 0x7f) {
+			char buf[3];
+			sprintf(buf, "%02x", ch & 0xff);
+			output++ = '\\'; output++ = 'x'; output++ = buf[0]; output++ = buf[1];
+		}
+		else {
+			output++ = ch;
+		}
+	}
+
+	return output;
+}
+
+std::string headers_to_string(
+		const ioremap::swarm::http_headers& headers,
+		const std::vector<std::string>& log_headers,
+		int quote = '"'
+	)
+{
+	typedef std::back_insert_iterator<std::string> output_type;
+
+	std::string headers_string;
+	output_type output(headers_string);
+
+	output++ = '{';
+
+	bool first_header = true;
+	for (const auto& log_header: log_headers) {
+		if (headers.has(log_header)) {
+			const auto& header_value = *headers.get(log_header);
+
+			if (!first_header) {
+				output++ = ','; output++ = ' ';
+			}
+			else {
+				first_header = false;
+			}
+
+			output++ = quote;
+			output = std::copy(log_header.begin(), log_header.end(), output);
+			output++ = ':'; output++ = ' ';
+			output = escape(header_value.begin(), header_value.end(), output, quote);
+			output++ = quote;
+		}
+	}
+
+	output++ = '}';
```

Comment:
Please run performance test.
It is needed not to stop this from being merged, but to understand what the performance implications are.

If one has these simple numbers - plain HTTP performance without header dump and the same test with one-two-something headers dump - it is possible to understand and predict how many (and which) headers has to be dumped to only decrease performance in allowed range.